### ISSUE TITLE
Regression since gradle 4.10 when using tasks created by rules

### ISSUE THREAD

**Issue by jnizet** [gradle/kotlin-dsl#1097]
I don't know if it's a Kotlin DSL issue or a gradle issue, so I'm posting it here. Please tell if I should file it on the gradle project.

The actual issue happened when using the [gradle node plugin](https://plugins.gradle.org/plugin/com.moowork.node), but I managed to reproduce it with a very simple test case:

```
tasks.addRule("Pattern: 'foo_<command>': prints the command.") {
    if (startsWith("foo_")) {
        tasks.create(this) {
            doLast {
                println(substring(4))
            }
        }
    }
}

tasks {
    val foo_hello by getting

    create("hello") {
        dependsOn(foo_hello)
    }
}
```

### Expected Behavior
Executing the hello task should work fine and print hello.

### Current Behavior
 - With gradle 4.9: it works as expected
 - With gradle 4.10: the configuration fails with the following error:

```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/jb/tmp/gradle-rule-bug/build.gradle.kts' line: 12

* What went wrong:
Task with name 'foo_hello' not found in root project 'gradle-rule-bug'.
```

### Context
I'm trying to upgrade a project with a [working build script](https://github.com/Ninja-Squad/globe42/blob/master/frontend/build.gradle.kts) to gradle 4.10.

**Comment by oehme** [comment]
That's a bug. @bamboo can you have a look at this?

**Comment by jnizet** [comment]
Note: if I try to avoid relying on rules by rewriting my build script as follows, it's properly configured and works fine with gradle 4.9, but fails with gradle 4.10 with the following error:

```

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/jb/projects/globe42/frontend/build.gradle.kts' line: 69

* What went wrong:
Could not create task ':frontend:test'.
> Could not create task ':frontend:yTest'.
   > Could not create task of type 'YarnTask'.
      > Project#afterEvaluate(Closure) on project ':frontend' cannot be executed in the current context.
```

I'm not sure if it's the same issue or another one.

```
import com.moowork.gradle.node.yarn.YarnTask

plugins {
    base
    id("com.moowork.node") version "1.2.0"
}

node {
    version = "8.11.3"
    npmVersion = "6.1.0"
    yarnVersion = "1.7.0"
    download = true
}

tasks {
    getByName("npmInstall") {
        enabled = false
    }

    val yInstall by creating(YarnTask::class) {
        args = listOf("install")
        inputs.file("package.json")
        inputs.file("yarn.lock")
        outputs.dir("node_modules")
    }

    val deleteMomentLocales by creating(Delete::class) {
        dependsOn(yInstall)
        delete(fileTree("${projectDir}/node_modules/moment/locale") {
            include("*.js")
        })
    }

    val prepare by creating {
        dependsOn(deleteMomentLocales)
    }

    val yBuild by creating(YarnTask::class) {
        args = listOf("build")
        dependsOn(prepare)
        inputs.dir("src")
        outputs.dir("dist")
    }

    val yTest by creating(YarnTask::class) {
        args = listOf("test")
        dependsOn(prepare)
        inputs.dir("src")
        outputs.dir("coverage")
    }

    val test by creating {
        dependsOn(yTest)
    }

    getByName("check") {
        dependsOn(test)
    }

    getByName("assemble") {
        dependsOn(yBuild)
    }

    getByName("clean") {
        dependsOn("cleanYBuild")
        dependsOn("cleanYTest")
    }
}
```

**Comment by oehme** [comment]
I guess the YarnTask calls afterEvaluate in its constructor, which is a really bad idea. Tasks should not call into the project, they should be configured from the outside by a plugin.

**Comment by jnizet** [comment]
Indeed: https://github.com/srs/gradle-node-plugin/blob/master/src/main/groovy/com/moowork/gradle/node/yarn/YarnTask.groovy. 

But still, this worked fine in 4.9, and it doesn't anymore in 4.10.

At some point, I guess this plugin (which I just use, but didn't write) should be forked and rewritten. I'd expect such failures when upgrading to 5.0, but not from 4.9 to 4.10.

**Comment by lacasseio** [comment]
I will bring this up as a possible regression to fix in our planning meeting today. Thanks for reporting this issue.

**Comment by lacasseio** [comment]
Isn't `val yTest by creating(YarnTask::class)` would eagerly create the task? The workaround is to for Kotlin to create the task using the eager API as opposed to the now default lazy API.

**Comment by jnizet** [comment]
Yes, this would eagerly create the task. But, AFAIK, so does `val yarn_install by getting`: it would ask the container to get a task eagerly, and the rule would eagerly create it (see https://github.com/srs/gradle-node-plugin/blob/master/src/main/groovy/com/moowork/gradle/node/NodePlugin.groovy#L100-L103)

**Comment by big-guy** [comment]
In 4.9, `getting` called `getByName`, which worked with tasks created via rules.
In 4.10, `getting` changed to calling `named`, which didn't work with tasks created via rules at the time (https://github.com/gradle/gradle-native/issues/718).
https://github.com/gradle/kotlin-dsl/blob/75ad0e760f9c6fe2d3f0728f60837a0b22184566/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt#L362-L365

`named` is fixed in master and would be in the next major release.

@eskatos not sure what you want to do about this.  We could try cherry-picking the fix to release if it's not tied up into too many other changes.

**Comment by jnizet** [comment]
That looks like a bug then: the migration guide (https://guides.gradle.org/migrating-build-logic-from-groovy-to-kotlin/#configuring-tasks) says that getting() gets a task eagerly, and that existing gets a task lazily.

**Comment by jnizet** [comment]
That said, shouldn't the tasks created by rules be usable as a dependency of other tasks whatever the strategy (getting vs. existing) is?

**Comment by oehme** [comment]
Yes, that's a bug in lazy task creation I'd say.

**Comment by eskatos** [comment]
@big-guy, if cherry-picking the existing `gradle/gradle` fix on `release` for 4.10.1 is possible that'd be the best solution

**Comment by big-guy** [comment]
@eskatos I opened https://github.com/gradle/gradle/pull/6706

**Comment by eskatos** [comment]
Thank you @big-guy !

**Comment by eskatos** [comment]
Closing as delivered in gradle/gradle#6706, thanks again @big-guy!

**Comment by JLLeitschuh** [comment]
Great find @jnizet!

**Comment by wolfs** [comment]
Fixed in 4.10 via https://github.com/gradle/gradle/pull/6706.
